File: solutions/clojure/yacht/1/src/yacht.clj
(ns yacht
  (:require [clojure.string :as str]))

(defmulti dice-score (fn [arg1 _] arg1))

(defmethod dice-score :yacht [_ freq]
  (when (= 1 (count freq)) 50))

(defn- sum-dice [freq]
  (->> freq
       (map (partial apply *))
       (apply +)))

(defn- with-count-sorted [freq]
  (let [[[n1 q1] [n2 q2]] (sort-by second freq)]
    [(count freq) n1 q1 n2 q2]))

(defmethod dice-score :full-house [_ freq]
  (let [[n _ q1 _ q2] (with-count-sorted freq)]
    (and (= 2 n)
         (= 2 q1)
         (= 3 q2)
         (sum-dice freq))))

(defmethod dice-score :four-of-a-kind [_ freq]
  (let [[total n1 q1 n2 q2] (with-count-sorted freq)
        n (or n2 n1)]
    (and (<= 1 total 2)
         (or (= 5 q1) (= 4 q2))
         (* n 4))))


(defn- set-check
  [expected freq]
  (when (= expected (set (keys freq))) 30))

(defmethod dice-score :little-straight [_ freq]
  (set-check #{1 2 3 4 5} freq))


(defmethod dice-score :big-straight [_ freq]
  (set-check #{2 3 4 5 6} freq))


(defmethod dice-score :choice [_ freq]
  (sum-dice freq))

(def ->dice
  {:ones 1 :twos 2 :threes 3 :fours 4 :fives 5 :sixes 6})

(defmethod dice-score :default [k freq]
  (let [n (->dice k)]
    (* n (freq n 0))))

(defn ->kebab [s]
  (keyword (str/replace s " " "-")))

(defn score [dice category]
  (or
   (dice-score (->kebab category) (frequencies dice))
   0))

File: solutions/scala/yacht/1/src/main/scala/Yacht.scala
object Yacht {
  private type Die = Int
  private type Dice = List[Die]

  private val LittleStraight: Dice = List(1, 2, 3, 4, 5)
  private val BigStraight: Dice = List(2, 3, 4, 5, 6)

  def score(dice: Dice, category: String): Int =
    val frequencies = dice.groupBy(identity).view.mapValues(_.length).toMap

    category match {
      case "ones" => sumOf(dice, 1)
      case "twos" => sumOf(dice, 2)
      case "threes" => sumOf(dice, 3)
      case "fours" => sumOf(dice, 4)
      case "fives" => sumOf(dice, 5)
      case "sixes" => sumOf(dice, 6)
      case "full house" => fullHouseScore(dice, frequencies)
      case "four of a kind" => fourOfAKindScore(frequencies)
      case "little straight" => straightScore(dice, LittleStraight)
      case "big straight" => straightScore(dice, BigStraight)
      case "choice" => dice.sum
      case "yacht" => if frequencies.size == 1 then 50 else 0
      case _ => 0
    }

  private def sumOf(dice: Dice, number: Die): Int =
    dice.filter(_ == number).sum

  private def fullHouseScore(dice: Dice, frequencies: Map[Die, Int]): Int =
    if frequencies.values.toSet == Set(2, 3) then dice.sum else 0

  private def fourOfAKindScore(frequencies: Map[Die, Int]): Int =
    frequencies
      .find { case (_, count) => count >= 4 }
      .fold(0) { case (number, _) => number * 4 }

  private def straightScore(dice: Dice, straight: Dice): Int =
    if dice.sorted == straight then 30 else 0
}
